### PR TITLE
Adds determinism to REPRO for better error checking & reproducibility

### PR DESCRIPTION
**Pull request by jefflarkin**
Makes gradient and energy calculations deterministic. 

Merged from https://github.com/scottlegrand/AutoDock-GPU/tree/relicensing-deterministic

### CODE REVIEW

File: host/inc/miscellaneous.h [hunk 1]
```diff
@@ -100,57 +104,44 @@ int strincmp(const char*, const char*, int);
...
-	}
-
-	float random_float(){
-		float result = 0.0f;
-		// Ensure random number is between 0 and 1
-		while (result<=0.0f || result>=1.0f)
-			result = (float)random_uint()/4294967295.0f;
-		return result;
-	}
+        //The function generates random numbers with a linear congruential generator, using Visual C++ generator constants.
```
Comment (atillack) [inline host/inc/miscellaneous.h]: Let's please not regress to linear congruential ;-)

File: host/inc/miscellaneous.h [hunk 2]
```diff
@@ -100,57 +104,44 @@ int strincmp(const char*, const char*, int);
...
+        unsigned int state;
+        
+        void Init(unsigned int seed)
+        {
+            state = seed;
+        }
+        
+
+        public:
+        LocalRNG(unsigned int seed)
```
Comment (atillack) [inline host/inc/miscellaneous.h]: That's the only additional function we should add to LocalRNG - with the argument being a uint32_t

File: host/src/performdocking.cpp.Cuda [hunk 3]
```diff
@@ -408,7 +408,9 @@ filled with clock() */
 	size_prng_seeds = blocksPerGridForEachEntity * threadsPerBlock * sizeof(unsigned int);
 	cpu_prng_seeds = (unsigned int*) malloc(size_prng_seeds);
 
-	LocalRNG r;
+	LocalRNG r(mypars->seed);
```
Comment (atillack) [inline host/src/performdocking.cpp.Cuda:411]: Please add to performdocking.cpp.OpenCL as well ...

File: host/inc/getparameters.h [hunk 4]
```diff
@@ -50,6 +50,7 @@ typedef struct
 typedef struct
 //Struct which contains the docking parameters (partly parameters for fpga)
 {
+    unsigned int seed;
```
Comment (atillack) [inline host/inc/getparameters.h:53]: please switch to uint32_t

**Comment by jefflarkin** [comment]
@atillack Ready for some further testing. Keeping this as a draft while we compare the results.

**Comment by atillack** [comment]
Nice AD speedup - but there's a couple of bugs in the calculation ... Uploading a fix (if I still have write access) soon :-)

**Comment by jefflarkin** [comment]
Moving from draft to get further reviews.

**Comment by diogomart** [comment]
Convergence tests failing as search performance is very poor. Values in the axis are the number of evals (in thousands) needed to achieve 50% success rate.
c1355120-ADRHO_vs_7c6652a-cuda-128wi_ad

**Comment by atillack** [comment]
@diogomart Thank you! Now on to fixing it :-)

**Comment by atillack** [comment]
@diogomart Please test again ... The E50 regression should be fixed now.

**Comment by atillack** [comment]
@jefflarkin Just a quick note, initial tests look good now and a more extensive test is on the way.

**Comment by diogomart** [comment]
Good news, convergence got a little better, some systems now need 4x fewer evaluations to get the 50% success rate.
e48_vs_ffd3057-cuda-128wi_ad

**Comment by atillack** [comment]
Thank you, @diogomart :-) Just to reiterate and add a bit more data to it. After the fixes, we now have better convergence and retain the nice speedup on top:
This is for Diogo's 42 on our Titan V node with `NUMWI=128`, default nr of evals, `nr_runs=100`, and `lsrat=100.0` and the current develop branch:
> # stats:        42                       105000000       36  /  42 good                          1024.28 s      90.15 s
This is the current state of the PR (which is also what Diogo used):
> # stats:        42                       105000000       39  /  42 good                           781.01 s      89.70 s

Here is the same thing with `nr_runs=10`:
Current:
> # stats:        42                       105000000       35  /  42 good                           166.27 s      36.84 s
The PR currently:
> # stats:        42                       105000000       36  /  42 good                           109.24 s      36.77 s
All runs were run back-to-back on the same node.

So, 20%-50% speedup and better E50 convergence ... What's not to like :-)

Of course, now the big task is to make sure this didn't break anything we haven't looked at yet ... But I am cautiously optimistic ...

**Review by diogomart — APPROVED** [review]

**Comment by atillack** [comment]
@diogomart Please test the OpenCL implementation. It looks OK on my end.

**Comment by diogomart** [comment]
OpenCL implementation looking good as well :-)
e48_vs_6db4c2e-ocl-128wi_ad

**Comment by atillack** [comment]
Thank you @diogomart!

Here are the performance numbers for OpenCL of Diogo's 42 run identically to the Cuda ones up above, `nr_runs=100`:
> # stats:        42                       105000000       39  /  42 good                           483.01 s      90.04 s
And `nr_runs=10`
> # stats:        42                       105000000       35  /  42 good                            74.09 s      38.84 s

**Comment by atillack** [comment]
@jefflarkin @diogomart @jeeberhardt I think this successfully makes this code path viable.  As this is a major (and very promising) change compared to the original one we now need to make a decision how to move forward.

My suggestion is this:
- we should convince ourself with some really though systems that the new code path remains successful, and
- keep the original code path around in a define, so if there is the odd system where it might have downsides the old one could be used

What do you think?

**Comment by diogomart** [comment]
Sounds good to me!

**Comment by jefflarkin** [comment]
> @jefflarkin @diogomart @jeeberhardt I think this successfully makes this code path viable. As this is a major (and very promising) change compared to the original one we now need to make a decision how to move forward.
> 
> My suggestion is this:
> 
> * we should convince ourself with some really though systems that the new code path remains successful, and
> * keep the original code path around in a define, so if there is the odd system where it might have downsides the old one could be used
> 
> What do you think?

In my opinion, because this is all in git, it's trivial for someone who wishes to use the old code path to restore to that point in history and if it's really important to them to backport future enhancements.

**Comment by atillack** [comment]
@jefflarkin Because it is easy to do at this point I added the float gradients as a fallback code path for debugging. I left the genotype gradients as-is. This is in between the current (soon old) code and this PR and I think can serve as a good sanity check for potential future issues.

@diogomart @jeeberhardt From my end, I am happy with this PR as-is. What do you think?

**Review by jeeberhardt — APPROVED** [review]

**Review by diogomart — APPROVED** [review]

**Pull request merged**